### PR TITLE
#2882 - PT  Award Configuration SBSD

### PR DESCRIPTION
**Pull request by guru-aot**
Updated PartTime assessment gateway bpmn Award configuration SBSD to consider the programYearFullTimeSBSD + ProgramYearTotalPartTimeSBSD while generating the finalProvincialAwardNetSBSDAmount.

image

image

E2E created:
parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts
parttime-assessment-2022-2023-awards-amount-SBSD.e2e-spec.ts
parttime-assessment-2023-2024-awards-amount-SBSD.e2e-spec.ts

e2e updated:
parttime-assessment-2021-2022-eligibility-SBSD.e2e-spec.ts
parttime-assessment-2022-2023-eligibility-SBSD.e2e-spec.ts
parttime-assessment-2023-2024-eligibility-SBSD.e2e-spec.ts

Note:

Fixed the bug that generates null instead of 0 when calculating federalAwardNetCSGDAmount, federalAwardNetBCAGAmount. AC's are updated in the respective tickets.

### CODE REVIEW

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn [hunk 1]
```diff
@@ -770,13 +726,67 @@
...
+      <bpmn:outgoing>Flow_10dvc27</bpmn:outgoing>
+    </bpmn:intermediateThrowEvent>
+    <bpmn:exclusiveGateway id="Gateway_1wo2y47">
+      <bpmn:incoming>Flow_121hd80</bpmn:incoming>
+      <bpmn:outgoing>Flow_0x4ck3r</bpmn:outgoing>
+      <bpmn:outgoing>Flow_1kzoimw</bpmn:outgoing>
+    </bpmn:exclusiveGateway>
+    <bpmn:sequenceFlow id="Flow_01gq993" sourceRef="Event_19hr2p7" targetRef="Event_0irnnvd" />
+    <bpmn:sequenceFlow id="Flow_121hd80" sourceRef="Event_0irnnvd" targetRef="Gateway_1wo2y47" />
+    <bpmn:intermediateThrowEvent id="Event_0irnnvd" name="Program Year Total both FT and PT">
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn:781]: I would say to move this node before CSGP calculation as CSGP is also across FT and PT and this event will be a potential place to add program year total CSGP.

image

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      PROGRAM_YEAR,
+      assessmentConsolidatedData,
+    );
+
+    // Assert
+    expect(calculatedAssessment.variables.programYearTotalSBSD).toBe(90);
+    expect(
+      calculatedAssessment.variables.dmnPartTimeAwardAllowableLimits
+        .limitAwardSBSD40AndUpCourseLoadAmount,
+    ).toBe(800);
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts:25]: I would say that we shouldn't assert dmn value.

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    // Act
+    const calculatedAssessment = await executePartTimeAssessmentForProgramYear(
+      PROGRAM_YEAR,
+      assessmentConsolidatedData,
+    );
+
+    // Assert
+    expect(calculatedAssessment.variables.programYearTotalSBSD).toBe(90);
+    expect(
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts:49]: https://github.com/bcgov/SIMS/pull/2920/files#r1513170474

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      PROGRAM_YEAR,
+      assessmentConsolidatedData,
+    );
+
+    // Assert
+    expect(calculatedAssessment.variables.awardEligibilitySBSD).toBe(true);
+    expect(calculatedAssessment.variables.programYearTotalSBSD).toBe(1100);
+    expect(
+      calculatedAssessment.variables.dmnPartTimeAwardAllowableLimits
+        .limitAwardSBSD40AndUpCourseLoadAmount,
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts:80]: https://github.com/bcgov/SIMS/pull/2920/files#r1513170474

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    assessmentConsolidatedData.programYearTotalPartTimeSBSD = 600;
+
+    // Act
+    const calculatedAssessment = await executePartTimeAssessmentForProgramYear(
+      PROGRAM_YEAR,
+      assessmentConsolidatedData,
+    );
+
+    // Assert
+    expect(calculatedAssessment.variables.awardEligibilitySBSD).toBe(true);
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts:76]: This condition is expected to be true for all the above test scenarios as well. I am not sure if we need to assert to make sure here.

Btw it is not a blocker.

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/eligibility/parttime-assessment-2021-2022-eligibility-SBSD.e2e-spec.ts [hunk 6]
```diff
@@ -35,7 +35,7 @@ describe(`E2E Test Workflow parttime-assessment-${PROGRAM_YEAR}-eligibility-SBSD
         expect(calculatedAssessment.variables.awardEligibilitySBSD).toBe(true);
         expect(
           calculatedAssessment.variables.finalProvincialAwardNetSBSDAmount,
-        ).toBeGreaterThan(0);
+        ).toBe(710);
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/test/2021-2022/parttime-assessment/eligibility/parttime-assessment-2021-2022-eligibility-SBSD.e2e-spec.ts:38]: I would say award eligibility must not check for specific numbers

File: sources/packages/backend/workflow/test/2023-2024/parttime-assessment/awards-amounts/parttime-assessment-2023-2024-awards-amount-SBSD.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      calculatedAssessment.variables.dmnPartTimeAwardAllowableLimits
+        .limitAwardSBSD40AndUpCourseLoadAmount,
+    ).toBe(800);
+    expect(calculatedAssessment.variables.provincialAwardNetSBSDAmount).toBe(0);
+    expect(
+      calculatedAssessment.variables.finalProvincialAwardNetSBSDAmount,
+    ).toBe(calculatedAssessment.variables.provincialAwardNetSBSDAmount);
+  });
+
+  it("Should determine provincialAwardNetSBSDAmount as zero when awardEligibilitySBSD is false", async () => {
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/test/2023-2024/parttime-assessment/awards-amounts/parttime-assessment-2023-2024-awards-amount-SBSD.e2e-spec.ts:88]: This test is checking award eligibility which seems to be redundant with existing award eligibility tests.

**Comment by dheepak-aot** [comment]
Good Job @guru-aot . Please have a look at the comments.

**Review by sh16011993 — APPROVED** [review]
LGTM @guru-aot 👍 Nice work

File: sources/packages/backend/workflow/test/2023-2024/parttime-assessment/awards-amounts/parttime-assessment-2023-2024-awards-amount-SBSD.e2e-spec.ts (hunk 7 above)
Comment (guru-aot) [reply sources/packages/backend/workflow/test/2023-2024/parttime-assessment/awards-amounts/parttime-assessment-2023-2024-awards-amount-SBSD.e2e-spec.ts:76]: Added to cover additional tests related to false scenario awards calculation

**Review by lewischen-aot — APPROVED** [review]
LGTM @guru-aot Nice work 👍

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn [hunk 8]
```diff
@@ -770,13 +703,88 @@
...
+    <bpmn:intermediateThrowEvent id="Event_14d0375" name="calculate SBSD">
+      <bpmn:extensionElements>
+        <zeebe:ioMapping>
+          <zeebe:output source="=if awardEligibilitySBSD = true&#10;then&#10;max(dmnPartTimeAwardAllowableLimits.limitAwardSBSD40AndUpCourseLoadAmount - programYearTotalSBSD,0)&#10;else&#10;  0" target="provincialAwardNetSBSDAmount" />
+        </zeebe:ioMapping>
+      </bpmn:extensionElements>
+      <bpmn:incoming>Flow_1kzoimw</bpmn:incoming>
+      <bpmn:outgoing>Flow_10dvc27</bpmn:outgoing>
+    </bpmn:intermediateThrowEvent>
+    <bpmn:intermediateThrowEvent id="Event_0irnnvd" name="Program Year Total both FT and PT">
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn:758]: There is minor overlap on descriptions. 

image

File: sources/packages/backend/workflow/test/2023-2024/parttime-assessment/awards-amounts/parttime-assessment-2023-2024-awards-amount-SBSD.e2e-spec.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/test/2023-2024/parttime-assessment/awards-amounts/parttime-assessment-2023-2024-awards-amount-SBSD.e2e-spec.ts:76]: I would still recommend to add this test to award eligibility, but considering it as not a blocker.

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
+import { PROGRAM_YEAR } from "../../constants/program-year.constants";
```
Comment (dheepak-aot) [inline sources/packages/backend/workflow/test/2021-2022/parttime-assessment/awards-amounts/parttime-assessment-2021-2022-awards-amount-SBSD.e2e-spec.ts:1]: If we all devs agree, we must remove the program year from the file names as per https://github.com/bcgov/SIMS/pull/2923#discussion_r1513553502

**Comment by dheepak-aot** [comment]
Thanks for making the changes @guru-aot . Added 2 more comments. PR almost looks good to go.

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes @guru-aot . Looks Good.

**Pull request merged**